src/main/java/com/rackspacecloud/metrics/ingestionservice/influxdb/GCLineProtocolBackupService.java

```diff
@@ -0,0 +1,133 @@
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import java.io.IOException;
+import java.nio.channels.Channels;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.GZIPOutputStream;
+
+@Service
+@EnableConfigurationProperties(BackupProperties.class)
+public class GCLineProtocolBackupService implements LineProtocolBackupService {
+
+    // Reusable thread-safe date-time formatter for files going in a bucket
+    private static final DateTimeFormatter dateBucketFormat = DateTimeFormatter.ofPattern("yyyyMMdd")
+            .withZone(ZoneOffset.UTC);
+    private static final Logger LOGGER = LoggerFactory.getLogger(GCLineProtocolBackupService.class);
+    // The underlying blob storage. Dev/test will uses in-memory storage
+    private final Storage storage;
+    // The bucket where blobs will go
+    private final String cloudOutputBucket;
+
+    // Internal methods of this class attempt to talk to the Cache proxy
+    // instead of the GCLineProtocolBackupService object
+    @Resource
+    private LineProtocolBackupService self;
+
+    // To autowire Storage in prod we'll need to get an account and point
+    // GOOGLE_APPLICATION_CREDENTIALS to the json creds file
+    // for a service account
+    // as per https://www.baeldung.com/java-google-cloud-storage
+    @Autowired
+    public GCLineProtocolBackupService(Storage storage, BackupProperties backupProperties) {
+        Assert.notNull(storage, "Storage must not be null");
+        Assert.notNull(backupProperties, "The backup properties must not be null");
+        // TODO: remove gcs prefix to make abstraction cleaner
+        Assert.notNull(backupProperties.getGcsBackupBucket(), "The output bucket must not be null");
+        this.storage = storage;
+        this.cloudOutputBucket = backupProperties.getGcsBackupBucket();
+    }
+
+    /**
+     * We want data to be stored in the following format: /[profile-specific
+     * bucket]/[instance name]/[database name]/[retention policy
+     * name]/[YYYYMMDD]/[UUID of file].gz
+     *
+     * File is line-protocol, gzipped, size may require some testing
+     *
+     * @param location the bucket location (directory) of the file we are writing to
+     * @return the cached reference to the gzip output stream for that blob
+     */
+    @Cacheable(cacheNames = "lineProtocolBackupWriter")
+    public GZIPOutputStream getBackupStream(String location) throws IOException {
+        Assert.notNull(location, "location must not be null");
```

Comment:
Can `location` be an empty string? If not, then we should add check for that as well.

Reply:
Oh yeah. Could use `Assert.hasText`

https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/util/Assert.html#hasText-java.lang.String-java.lang.String-